### PR TITLE
Move configs to new Unit Parameters DB and Max HP limits

### PR DESCRIPTION
**Pull request by guilherme-gm**
### Pull Request Prelude
- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
This PR creates a new db file called Unit Parameters DB and moves a few battle configs to it:
- NaturalHealWeightRate
- MaxASPD
- MaxHP
- MaxStats

The reason to move those settings is that:
1. We will need different values between RE and Pre-RE (e.g. MaxHP, NaturalHealWeightRate)
2. We already have more than one different setting for it due to job differencs (MaxASPD, MaxStats)

Thus having this new file allows us to link jobs to parameter groups for each server type.

Finally, a small change but an actual new content: This PR also adds the maximum HP based on users level. A change introduced in kRO's episode 15.1

Huge thanks to Haru and 4144 for the discussions on this :)

**Issues addressed:** MaxHP limits item from #243 

### For users upgrading their Hercules copy
(I plan to also post this on herc board once the PR gets in a release)

There are 3 changes introduced in this PR that may affect you.

#### 1. job_db.conf update
If you are using Hercules' default db folder (`db/pre-re/job_db.conf` or `db/re/job_db.conf`) this should get merged automatically.

For custom folders, please check the original Hercules' `job_db.conf` and replicate the `ParametersGroup` of each job. Using the same values as in Hercules' original file will make the other 2 changes easier to migrate.

#### 2. You must have the new `unit_parameters_db.conf` db file
If you are using Hercules' default db folder (`db/pre-re/` or `db/re/`) this should get merged automatically.

For custom folders, please copy `db/(pre-re|re)/unit_parameters_db.conf` to your own db folder (based on your server type).

Those will already come with the default configs and groups that you needed in step 1.

#### 3. Migrate configs from `conf/battle/player.conf`
If you have customized any of the configs changed in this PR, you may have to update your `unit_parameters_db.conf` with the new values. Below you will find each battle config and how they map to the new `unit_parameters_db.conf`.

- the first column mentions the existing battle conf name
- the second column mentions the field in unit parameters db equivalent to this config (Note: Also look the 3rd column)
- the third column mentions for which group(s) this config applies (for example: a 3rd job-only config will affect only 3rd job groups)
- the fourth column gives you some notes, like whether you can remove or not this config and why.

| player.conf | Parameter Field | Parameter Group | Notes |
| ----------- | --------------- | ------------- | -------- |
| `natural_heal_weight_rate` | `NaturalHealWeightRate` | `Base` | This config may be removed from your player.conf, it is no longer used. |
| `max_aspd` | `MaxASPD` |`Base` | This config should be kept in your player.conf, since it is still used for homunculus and as fallback for players |
| `max_third_aspd` | `MaxASPD` | `ThirdClasses`<br/>`SuperNovice` | This config may be removed from your player.conf, it is no longer used. |
| `max_hp` | `MaxHP` | `Base` | This config may be removed from your player.conf, it is no longer used. You may be interested in the level grouping (check docs) |
| `max_parameter` | `MaxStats` | `Base` | This config should be kept in your player.conf, since it is still used for some skills/npc logic |
| `max_third_parameter` | `MaxStats` | `ThirdClasses` | This config should be kept in your player.conf, since it is still used for some skills/npc logic |
| `max_extended_parameter` | `MaxStats` | `ExtendedClasses` | This config may be removed from your player.conf, it is no longer used. |
| `max_summoner_parameter` | `MaxStats` | `Summoner` | This config may be removed from your player.conf, it is no longer used. |
| `max_baby_parameter` | `MaxStats` | `BabyBase` | This config may be removed from your player.conf, it is no longer used. |
| `max_baby_third_parameter` | `MaxStats` | `BabyThirdClasses` | This config may be removed from your player.conf, it is no longer used. |

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

File: src/map/battle.c [hunk 1]
```diff
@@ -7499,18 +7499,14 @@ static const struct config_data_old battle_data[] = {
 	{ "natural_healsp_interval",            &battle_config.natural_healsp_interval,         8000,   NATURAL_HEAL_INTERVAL, INT_MAX, },
 	{ "natural_heal_cap",                   &battle_config.natural_heal_cap,                1000,   1,      INT_MAX,        },
 	{ "natural_heal_skill_interval",        &battle_config.natural_heal_skill_interval,     10000,  NATURAL_HEAL_INTERVAL, INT_MAX, },
-	{ "natural_heal_weight_rate",           &battle_config.natural_heal_weight_rate,        50,     50,     101             },
```
[the remark this answers is not in the document]
Comment (guilherme-gm) [reply src/map/battle.c:7502]: sure, I like this idea.

Do you think a macro that does the same thing we do for `battle_configuration/traps_setting` is enough? Example:
```C
CHECK_DEPRECATED_BATTLECONF("natural_heal_weight_rate", "'natural_heal_weight_rate' was moved to unit_params_db.");
```

Maybe looping through an array of deprecated strings would be an option too.

Or you are thinking about something more elaborate that automatically detects invalid keys in the conf?

[the remark this answers is not in the document]
Comment (Asheraf) [reply src/map/battle.c:7502]: Either will be fine, just put them in a separate function so we can add to it easier in the future 👍

[the remark this answers is not in the document]
Comment (guilherme-gm) [reply src/map/battle.c:7502]: done! thanks for the suggestion

File: src/map/pc.h [hunk 2]
```diff
@@ -744,6 +744,7 @@ END_ZEROED_BLOCK;
 #define pc_isinvisible(sd)    ( (sd)->sc.option&OPTION_INVISIBLE )
 #define pc_is50overweight(sd) ( (sd)->weight*100 >= (sd)->max_weight*battle->bc->natural_heal_weight_rate )
 #define pc_is90overweight(sd) ( (sd)->weight*10 >= (sd)->max_weight*9 )
+#define pc_maxhp_cap(sd) ( status->get_maxhp_cap_entry(pc->class2idx(sd->status.class), sd->status.base_level)->value )
```
Comment (MishimaHaruna) [inline src/map/pc.h:748]:
```
#define pc_maxhp_cap(sd) ( status->get_maxhp_cap_entry(pc->class2idx((sd)->status.class), (sd)->status.base_level)->value )
```

Please add parentheses around the macro argument to be safe against unusual expressions

File: src/map/pc.h [hunk 3]
```diff
@@ -745,6 +745,7 @@ END_ZEROED_BLOCK;
 #define pc_is50overweight(sd) ( (sd)->weight*100 >= (sd)->max_weight*battle->bc->natural_heal_weight_rate )
 #define pc_is90overweight(sd) ( (sd)->weight*10 >= (sd)->max_weight*9 )
 #define pc_maxhp_cap(sd) ( status->get_maxhp_cap_entry(pc->class2idx(sd->status.class), sd->status.base_level)->value )
+#define pc_max_aspd(sd) ( status->dbs->unit_params[pc->class2idx(sd->status.class)]->max_aspd )
```
Comment (MishimaHaruna) [inline src/map/pc.h:749]:
```
#define pc_max_aspd(sd) ( status->dbs->unit_params[pc->class2idx((sd)->status.class)]->max_aspd )
```

Same as above

**Review by MishimaHaruna — CHANGES_REQUESTED** [review]

File: src/map/pc.h (hunk 2 above)
Comment (guilherme-gm) [reply src/map/pc.h]: oops. done. thank you!

**Review by MishimaHaruna — APPROVED** [review]

**Pull request merged**